### DISCUSSION TITLE
Regarding to `Node`

### DISCUSSION THREAD

**Discussion by minshao** [aicers/review-web#226] [Q&A] [answered]
* For a valid node, are we allowing `Node::name` to be empty string?
What about other string field, for example, `hostname` and `description` in `Node::settings`?

* When the user first insert the node, the node will have a `name` and `settings_draft`, and at this stage, `settings` is `None`. Does this mean that, at a later stage, it's possible and acceptable for a node to have `None` as `settings`?

Could you shed some light on the `Node` design @sehkone @sophie-cluml ?

**Comment by sehkone** [comment, marked as answer]
@minshao,

- `name` and `hostname` must not be empty, but `description` can be.
- `settings` is `None` only until the initial draft is applied to a node for the first time. When applying the draft, it must have valid values for the node. Once applied, `settings` will be replaced with `settings_draft`.

If you have any further questions about `Node`, feel free to ask anytime.

**Reply by minshao** [reply]
Thank you so much for the instant reply. 
In addition, say that a node initially has `Hog`,  will the node be able to hold `Piglet` or `Reconverge`? Or once the node is confirmed with `Hog`, the user can only update the values of `Hog`? I'm asking because I am planning to allow node to hold multiple agent (could be any combination of `hog`, `reconverge`, `piglet`, and potentially `crusher`, see the code below), the restrictions for update will be important for the modification. 
```Rust
enum Agent {
    Reconverge(Reconverge),
    Hog(Hog),
    Piglet(piglet),
}

struct Settings {
    ...
    agents: Vec<Agent>,
}
```

**Reply by sehkone** [reply]
@minshao,
Since a node represents a single real or virtual machine, the set of agents running on it may change over time.
